inferno-framework/smart-app-launch-test-kit
# FI-2247 backend services migration

# Summary
Migrating the Bulk Data Authorization tests into the SMART App Launch test kit.  Bulk Data test files were copied from the ONC (g)(10) Certification Test Kits, including [this PR from an external contributor](https://github.com/onc-healthit/onc-certification-g10-test-kit/pull/458), and now live under a Backend Services group for STU2.

## Updates From First Round of Feedback
- Test Structure
     - Backend services test group is now located after EHR launch and before Token Introspection group
     - Backend services test group includes SMART on FHIR Discovery tests as required by the IG, which will provide the token endpoint input
     - Individual tests now live in their own files
     - Eliminated redundant behavior between `ClientAssertionBuilder` and `AuthenticationRequestBuilder` 
          - Renamed `AuthenticationRequestBuilder` to `BackendServicesAuthorizationRequestBuilder` and had it use `ClientAssertionBuilder` 
- Inputs
     - Backend services tests now reuse an existing input for encryption method from standalone launch tests using confidential asymmetric client authorization, rather than having a separate, redundant input
     - Inferno Reference Server STU2 preset updated to include `backend_services_client_id`
          - Value was taken from ref server preset used on inferno.healthit.gov for bulk data authorization
          - All other inputs are either covered by default values or pre-existing presets
- Naming, Descriptions, & Requirements
     - Replaced all references to (g)(10) and/or bulk data with backend services in test names, inputs, and descriptions
     - Ensured all tests adhere to backend services STU 2.0.0 IG requirements and do not contain requirements specific to (g)(10) or bulk data 


# Testing Guidance
1. Run the test kit locally per the README instructions 
2. Navigate to [http://localhost:4567](http://localhost:4567/) in a web browser
3. Select SMART App Launch STU2
4. Select Inferno Reference Server as preset 
5. Select Group 3 Backend Services in the left-side menu, then Run Tests
6. Run the tests - all tests should pass
